### PR TITLE
Create get_fgpp.py

### PR DESCRIPTION
**Pull request by sebrink**
Initial script written by n00py, https://github.com/n00py/GetFGPP

Some sample output from my lab:

```
# poetry run NetExec ldap 172.16.14.200 -u user -p password -M get_fgpp          
SMB         172.16.14.200   445    COLOR-DC01       [*] Windows 10.0 Build 14393 x64 (name:COLOR-DC01) (domain:fmradio.local) (signing:True) (SMBv1:False)
LDAP        172.16.14.200   389    COLOR-DC01       [+] fmradio.local\user:password (Pwn3d!)
GET_FGPP    172.16.14.200   389    COLOR-DC01       [+] Attempting to enumerate policies...
GET_FGPP    172.16.14.200   389    COLOR-DC01       2 FGPP Objects found!
GET_FGPP    172.16.14.200   389    COLOR-DC01       
GET_FGPP    172.16.14.200   389    COLOR-DC01       [+] Attempting to enumerate objects with an applied policy...
GET_FGPP    172.16.14.200   389    COLOR-DC01       Object: CN=macheson,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       Applied Policy: 
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=second policy,CN=Password Settings Container,CN=System,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       
GET_FGPP    172.16.14.200   389    COLOR-DC01       Object: CN=breakglass01,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       Applied Policy: 
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=second policy,CN=Password Settings Container,CN=System,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=breakglass-fgpp,CN=Password Settings Container,CN=System,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       
GET_FGPP    172.16.14.200   389    COLOR-DC01       [+] Attempting to enumerate details...
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Name: breakglass-fgpp
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Length: 24
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password History Length: 
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Threshold: 0
GET_FGPP    172.16.14.200   389    COLOR-DC01       Observation Window: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Duration: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Complexity Enabled: TRUE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Age 1 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Maximum Password Age: 42 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Reversible Encryption: FALSE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Precedence: 1 (Lower is Higher Priority)
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Applies to: 
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=breakglass01,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Name: second policy
GET_FGPP    172.16.14.200   389    COLOR-DC01       Description: this policy is more of a suggestion tbh
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Length: 0
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password History Length: 
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Threshold: 23
GET_FGPP    172.16.14.200   389    COLOR-DC01       Observation Window: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Duration: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Complexity Enabled: TRUE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Age 1 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Maximum Password Age: 42 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Reversible Encryption: TRUE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Precedence: 2 (Lower is Higher Priority)
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Applies to: 
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=breakglass01,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=macheson,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01   
```

### CODE REVIEW

**Comment by mpgn** [comment]
Hello @sebrink what's the diff with the module PSO ? https://github.com/Pennyw0rth/NetExec/blob/72e42e9b75c9885573eecfc3888252eb93dfba9d/nxc/modules/pso.py

**Comment by sebrink** [comment]
@mpgn I actually was not aware of the PSO module, so these currently do the exact same thing. There appears to be a bug in the PSO module though that it can't display if there is a policy that applies to multiple objects, see the sample output below: (see the policy titled `second policy`)

get_fgpp:
```
# poetry run NetExec ldap 172.16.14.200 -u username -p password -M get_fgpp
SMB         172.16.14.200   445    COLOR-DC01       [*] Windows 10.0 Build 14393 x64 (name:COLOR-DC01) (domain:fmradio.local) (signing:True) (SMBv1:False)
LDAP        172.16.14.200   389    COLOR-DC01       [+] fmradio.local\rod:yewdall54 (Pwn3d!)
GET_FGPP    172.16.14.200   389    COLOR-DC01       [+] Attempting to enumerate policies...
GET_FGPP    172.16.14.200   389    COLOR-DC01       2 FGPP Objects found!
GET_FGPP    172.16.14.200   389    COLOR-DC01
GET_FGPP    172.16.14.200   389    COLOR-DC01       [+] Attempting to enumerate objects with an applied policy...
GET_FGPP    172.16.14.200   389    COLOR-DC01       Object: CN=macheson,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       Applied Policy:
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=second policy,CN=Password Settings Container,CN=System,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01
GET_FGPP    172.16.14.200   389    COLOR-DC01       Object: CN=breakglass01,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01       Applied Policy:
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=second policy,CN=Password Settings Container,CN=System,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=breakglass-fgpp,CN=Password Settings Container,CN=System,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01
GET_FGPP    172.16.14.200   389    COLOR-DC01       [+] Attempting to enumerate details...
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Name: breakglass-fgpp
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Length: 15
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password History Length:
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Threshold: 0
GET_FGPP    172.16.14.200   389    COLOR-DC01       Observation Window: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Duration: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Complexity Enabled: TRUE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Age 1 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Maximum Password Age: 42 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Reversible Encryption: FALSE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Precedence: 1 (Lower is Higher Priority)
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Applies to:
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=breakglass01,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Name: second policy
GET_FGPP    172.16.14.200   389    COLOR-DC01       Description: this policy is more of a suggestion tbh
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Length: 7
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password History Length:
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Threshold: 23
GET_FGPP    172.16.14.200   389    COLOR-DC01       Observation Window: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Lockout Duration: 0 days 0 hours 30 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Complexity Enabled: TRUE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Minimum Password Age 1 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Maximum Password Age: 42 days 0 hours 0 minutes 0 seconds
GET_FGPP    172.16.14.200   389    COLOR-DC01       Reversible Encryption: TRUE
GET_FGPP    172.16.14.200   389    COLOR-DC01       Precedence: 2 (Lower is Higher Priority)
GET_FGPP    172.16.14.200   389    COLOR-DC01       Policy Applies to:
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=breakglass01,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01           CN=macheson,CN=Users,DC=fmradio,DC=local
GET_FGPP    172.16.14.200   389    COLOR-DC01
```

pso:
```
# poetry run NetExec ldap 172.16.14.200 -u username -p password -M pso
SMB         172.16.14.200   445    COLOR-DC01       [*] Windows 10.0 Build 14393 x64 (name:COLOR-DC01) (domain:fmradio.local) (signing:True) (SMBv1:False)
LDAP        172.16.14.200   389    COLOR-DC01       [+] fmradio.local\username:password (Pwn3d!)
PSO         172.16.14.200   389    COLOR-DC01       [+] Password Settings Objects (PSO) found:
PSO         172.16.14.200   389    COLOR-DC01       cn: breakglass-fgpp
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordReversibleEncryptionEnabled: FALSE
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordSettingsPrecedence: 1
PSO         172.16.14.200   389    COLOR-DC01       msDS-MinimumPasswordLength: 15
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordHistoryLength: 24
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordComplexityEnabled: TRUE
PSO         172.16.14.200   389    COLOR-DC01       msDS-LockoutObservationWindow: 30 mins
PSO         172.16.14.200   389    COLOR-DC01       msDS-LockoutDuration: 30 mins
PSO         172.16.14.200   389    COLOR-DC01       msDS-LockoutThreshold: 0
PSO         172.16.14.200   389    COLOR-DC01       msDS-MinimumPasswordAge: 1 days
PSO         172.16.14.200   389    COLOR-DC01       msDS-MaximumPasswordAge: 42 days
PSO         172.16.14.200   389    COLOR-DC01       msDS-PSOAppliesTo: CN=breakglass01,CN=Users,DC=fmradio,DC=local
PSO         172.16.14.200   389    COLOR-DC01       -----
PSO         172.16.14.200   389    COLOR-DC01       cn: second policy
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordReversibleEncryptionEnabled: TRUE
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordSettingsPrecedence: 2
PSO         172.16.14.200   389    COLOR-DC01       msDS-MinimumPasswordLength: 7
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordHistoryLength: 0
PSO         172.16.14.200   389    COLOR-DC01       msDS-PasswordComplexityEnabled: TRUE
PSO         172.16.14.200   389    COLOR-DC01       msDS-LockoutObservationWindow: 30 mins
PSO         172.16.14.200   389    COLOR-DC01       msDS-LockoutDuration: 30 mins
PSO         172.16.14.200   389    COLOR-DC01       msDS-LockoutThreshold: 23
PSO         172.16.14.200   389    COLOR-DC01       msDS-MinimumPasswordAge: 1 days
PSO         172.16.14.200   389    COLOR-DC01       msDS-MaximumPasswordAge: 42 days
PSO         172.16.14.200   389    COLOR-DC01       msDS-PSOAppliesTo: CN=breakglass01,CN=Users,DC=fmradio,DC=local
PSO         172.16.14.200   389    COLOR-DC01       -----
```

If you'd like, I can make an update to the PSO module to fix the bug sometime tomorrow or it can potentially be replaced with the get_fgpp module, up to ya'll.

**Comment by mpgn** [comment]
Your output is better imo

**Comment by mpgn** [comment]
Maybe name your module pso and delete the old one since bugged anyway

**Comment by NeffIsBack** [comment]
Agreed, looks better.

@mpgn glad to see you here:)

**Comment by sebrink** [comment]
Alright it's updated to pso, modified the description and the output to say PSO rather than fgpp. As a note, the original pso module wasn't listed in the e2e_commands.txt.

**Comment by Marshall-Hallenbeck** [comment]
> Alright it's updated to pso, modified the description and the output to say PSO rather than fgpp. As a note, the original pso module wasn't listed in the e2e_commands.txt.

Thank you for adding it to the e2e tests :)
Could you update the single quotes to double to match our style? Thanks for the PR!

**Comment by sebrink** [comment]
@Marshall-Hallenbeck done!

**Comment by Marshall-Hallenbeck** [comment]
Changed to merge into `develop`
@sebrink there's a conflict with the poetry.lock if you can fix that

**Comment by sebrink** [comment]
Resolved the conflict, looks like running `poetry lock` installs greenlet v3.0.0 rather than greenlet v2.0.2 now using poetry 1.6.1. Happens with the `pyproject.toml` in main right now as well.

**Comment by sebrink** [comment]
That's almost definitely due to the LDAP query limit of 1000 objects. My lab has less than 1000 so I didn't test for that. I'll put some logic in to catch that. Need to figure out a good way to do it without just capping the module at the first 1000 entries, but nice catch @mpgn !

**Comment by mpgn** [comment]
Hum looking at the module, if a client has a pso on 7k user, the module will be unreadable

**Comment by mpgn** [comment]
Going after each object of the domain is overkill, imagine having 20k users

**Comment by mpgn** [comment]
Fixed the code for you @sebrink 
image

File: nxc/modules/pso.py [hunk 1]
```diff
@@ -1,109 +1,134 @@
 #!/usr/bin/env python3
 # -*- coding: utf-8 -*-
+# Toss this in /root/NetExec/nxc/modules
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/pso.py:3]: Left over testing comment?

File: nxc/modules/pso.py [hunk 2]
```diff
@@ -1,109 +1,134 @@
...
-                # We reached the sizeLimit, process the answers we have already and that's it. Until we implement
-                # paged queries
-                resp = e.getAnswers()
-                pass
-            else:
-                logging.debug(e)
-                return False
+        # Set some variables
+        self.__domain = connection.domain
+        self.__kdcHost = connection.hostname + "." + connection.domain
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/pso.py:33]: Use fstrings to concatenate strings here

File: nxc/modules/pso.py [hunk 3]
```diff
@@ -1,109 +1,134 @@
...
-                for field in self.pso_fields:
-                    if field in pso:
-                        value = self.convert_time_field(field, pso[field])
-                        context.log.highlight(u'{}: {}'.format(field, value))
-                context.log.highlight('-----')
+def base_creator(domain):
+    search_base = ""
+    base = domain.split(".")
+    for b in base:
+        search_base += "DC=" + b + ","
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/pso.py:128]: Use fstring for string concatenation

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]

File: nxc/modules/pso.py (hunk 1 above)
Comment (sebrink) [reply nxc/modules/pso.py]: yep, removed

**Comment by sebrink** [comment]
Thanks for the fix @mpgn!

**Comment by Marshall-Hallenbeck** [comment]
@sebrink my big refactor got merged in. Would you mind taking a look at the few conflicts?

**Comment by mpgn** [comment]
@sebrink any chance to fix the conflict to merge the pr ?

**Comment by NeffIsBack** [comment]
@mpgn fixed everything, could you give this another review?

**Comment by mpgn** [comment]
image

**Review by mpgn — APPROVED** [review]

**Comment by NeffIsBack** [comment]
@Marshall-Hallenbeck here are your format strings :)

**Review by Marshall-Hallenbeck — APPROVED** [review]

**Pull request merged**